packages/@lwc/engine-core/src/libs/signal-tracker/index.ts

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright (c) 2024, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+import { isFalse, isUndefined } from '@lwc/shared';
+import { Signal } from '@lwc/signals';
+import { logWarnOnce } from '../../shared/logger';
+
+/**
+ * This map keeps track of objects to signals. There is an assumption that the signal is strongly referenced
+ * on the object which allows the SignalTracker to be garbage collected along with the object.
+ */
+const TargetToSignalTrackerMap: WeakMap<Object, SignalTracker> = new WeakMap();
+
+function getSignalTracker(target: Object) {
+    let signalTracker = TargetToSignalTrackerMap.get(target);
+    if (isUndefined(signalTracker)) {
+        signalTracker = new SignalTracker();
+        TargetToSignalTrackerMap.set(target, signalTracker);
+    }
+    return signalTracker;
+}
+
+export function subscribeToSignal(
+    target: Object,
+    signal: Signal<unknown>,
+    update: CallbackFunction
+) {
+    const signalTracker = getSignalTracker(target);
+    if (isFalse(signalTracker.seen(signal))) {
+        signalTracker.subscribeToSignal(signal, update);
+    }
+}
+
+export function unsubscribeFromSignals(target: Object) {
+    if (TargetToSignalTrackerMap.has(target)) {
+        const signalTracker = getSignalTracker(target);
+        signalTracker.unsubscribeFromSignals();
+        signalTracker.reset();
+    }
+}
+
+type CallbackFunction = () => void;
+
+/**
+ * This class is used to keep track of the signals associated to a given object.
+ * It is used to prevent the LWC engine from subscribing duplicate callbacks multiple times
+ * to the same signal. Additionally, it keeps track of all signal unsubscribe callbacks, handles invoking
+ * them when necessary and discarding them.
+ */
+class SignalTracker {
+    private signalToUnsubscribeMap: Map<Signal<unknown>, CallbackFunction> = new Map();
+
+    seen(signal: Signal<unknown>) {
+        return this.signalToUnsubscribeMap.has(signal);
+    }
+
+    subscribeToSignal(signal: Signal<unknown>, update: CallbackFunction) {
+        try {
+            const unsubscribe = signal.subscribe(update);
+            if (typeof unsubscribe === 'function') {
+                this.signalToUnsubscribeMap.set(signal, unsubscribe);
+            }
+        } catch (err) {
+            logWarnOnce(
+                `Attempted to subscribe to an object that has the shape of a signal but received the following error: ${err}`
+            );
+        }
+    }
+
+    unsubscribeFromSignals() {
+        try {
+            this.signalToUnsubscribeMap.forEach((unsubscribe) => unsubscribe());
+        } catch (err) {
+            logWarnOnce(
+                `Attempted to call a signal's unsubscribe callback but received the following error: ${err}`
+            );
```

Comment:
This one seems a bit more serious to me. In this case, we might have landed here because a signal exposed a `value` and a `subscribe`, but not an `unsubscribe`. (I.e. you may have a signal implementation that is prone to memory leaks, since it doesn't implement `unsubscribe`.) I wonder if there is value in separating out the two cases and providing a better error message for each scenario.

Reply:
Elsewhere, we are checking that `typeof unsubscribe === 'function'`. So we know at this point that `unsubscribe` is invokable. However, similar to the other condition, this could occur if an error occurs _during_ unsubscribe, e.g. the signal wraps a websocket connection, `unsubscribe()` attempts to `ws.close()` and that fails for [some reason](https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/close#exceptions).

Same ask as earlier for preserving the error's stack.

Reply:
@nolanlawson In `SignalTracker.subscribeToSignal` we now check that the `subscribe` returns an `unsubscribe` function and only stores if it's a function.

For v1 I'll log the stack trace as @divmain suggested but I left a [todo](https://github.com/salesforce/lwc/blob/3c78f17484831774b00c494ab1e009a282f2a175/packages/%40lwc/engine-core/src/libs/signal-tracker/index.ts#L64-L65) (#3978) to revisit this in 252 once we figure out the use cases and know how we want to better handle the errors.